From d9701b824f5f381022ea8ef3db9b435784e63cd3 Mon Sep 17 00:00:00 2001
From: Jonathan Lemon <jonathan.lemon@gmail.com>
Date: Fri, 7 May 2021 13:59:18 -0700
Subject: [PATCH] spi-nor: Send soft-reset before probing flash.

On my FPGA board, after a powercycle, the first SPI command sent
to the flash chip always fails (and returns 0xff).  Most likely,
the FPGA load left things in a bad state.

Avoid this by sending a soft-reset to reinitialize/wakeup the chip
before sending a JEDEC identifcation request.

This allows things to work afer a powercycle, reboot sequence,
and a module driver unload/load cycle.

Signed-off-by: Jonathan Lemon <jonathan.lemon@gmail.com>
---
 drivers/mtd/spi-nor/core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/mtd/spi-nor/core.c b/drivers/mtd/spi-nor/core.c
index bd2c7717eb10..3531919fddce 100644
--- a/drivers/mtd/spi-nor/core.c
+++ b/drivers/mtd/spi-nor/core.c
@@ -3091,6 +3091,9 @@ int spi_nor_scan(struct spi_nor *nor, const char *name,
 	if (!nor->bouncebuf)
 		return -ENOMEM;
 
+	/* soft reset to wakeup/prepare chip */
+	spi_nor_soft_reset(nor);
+
 	info = spi_nor_get_flash_info(nor, name);
 	if (IS_ERR(info))
 		return PTR_ERR(info);
-- 
2.27.0

